easy-ui-react/src/TextField/TextField.tsx

```diff
@@ -0,0 +1,359 @@
+  }
+
+  const isPassword = type === "password";
+  const hasError = validationState === "invalid";
+  const showErrorText = hasError && errorText;
+  const showHelperText = !showErrorText && helperText;
+  const canUseIcon = !bothIconPropsDefined && !isPassword;
+  const hasStartIcon = canUseIcon && iconAtStart;
+  const hasEndIcon = canUseIcon && iconAtEnd;
+  const typeAdjustedForPasswordVisibility = isPassword && isVisible;
+
+  return (
+    <div className={classNames(styles.root)}>
+      <div className={styles.inputLabelContainer}>
+        <label
+          {...labelProps}
+          className={classNames(
+            styles.label,
+            isLabelVisuallyHidden && styles.labelHidden,
+          )}
+        >
+          {getLabelText(
+            emphasizedLabel,
+            hasError,
+            isLabelVisuallyHidden,
+            label,
+            size,
+          )}
+        </label>
+        <div className={styles.inputIconContainer}>
+          {hasStartIcon && getIcon(iconAtStart, true, size, isDisabled)}
+          <input
+            {...inputProps}
+            className={classNames(
+              styles.input,
+              isPassword && styles.passwordInput,
+              hasError && styles.errorInput,
+              hasStartIcon && styles.iconStartInput,
+              hasEndIcon && styles.iconEndInput,
+              styles[variationName("inputSize", size)],
+            )}
+            ref={ref}
+            type={typeAdjustedForPasswordVisibility ? "text" : type}
+            value={value}
+            required={isRequired}
+            disabled={isDisabled}
+            placeholder={placeholder}
+            autoFocus={autoFocus}
+            defaultValue={defaultValue}
+          />
+          {isPassword ? (
+            <UnstyledButton
+              className={classNames(
+                styles.passwordBtn,
+                hasError && styles.passwordBtnError,
+                styles[variationName("btnSize", size)],
+              )}
+              onPress={() => setIsVisible((prevVisibility) => !prevVisibility)}
+              isDisabled={isDisabled}
+            >
+              <Text visuallyHidden>password visibility</Text>
+              <Icon
+                symbol={isVisible ? VisibilityIcon : VisibilityOffIcon}
+                size={mapIconSize(size)}
+              />
+            </UnstyledButton>
+          ) : (
+            hasEndIcon && getIcon(iconAtEnd, false, size, isDisabled)
+          )}
+        </div>
+      </div>
+      {showHelperText && (
+        <div {...helperTextProps} className={styles.caption}>
+          <Text variant="caption" color="gray.resting">
+            {helperText}
+          </Text>
+        </div>
+      )}
+      {showErrorText && (
+        <div {...errorTextProps} className={styles.caption}>
+          <Text variant="caption" color="danger">
+            {errorText}
+          </Text>
+        </div>
+      )}
```

Comment:
**question:** it may be overkill so it's up to your discretion, but would it be worth creating a component that these each could use to eliminate the duplicities? `<TextFieldCaption />`

Reply:
created an `InputCaption` component; i think it will see some reuse in the `TextArea` component as well